Advanced SEO: Add Twitter Preview Component

vindl opened the pull request:
Closes #5971

Adds a Twitter preview component which can show summary card and summary card with large image for post object. Will eventually land on the editor preview pane.

**Summary card and Large image summary preview**

<img width="550" alt="screenshot from 2016-07-20 16-01-34" src="https://cloud.githubusercontent.com/assets/1182160/16989732/94a433a4-4e95-11e6-8cc9-828380218aa7.png">


**Testing**

This component should be tested with various description and title lengths in order to verify that text truncation is working properly. Description that this PR uses for testing can be changed in My Sites > Settings > SEO. Site title and tagline are used for card title (My Sites > Settings > General).

Previews do not currently pull the post content which would normally exist in the preview.

Test live: https://calypso.live/?branch=add/twitter-preview

vindl commented:
cc @dmsnell @rodrigoi

roundhill commented:
This also needs some fixes to make it responsive. For the Facebook component, we chose to make it overflow horizontally but allow scrolling. We could do the same here, or adjust the font sizes and image size to how it looks in the Twitter app when on mobile.

dmsnell commented on client/components/seo/twitter-preview/index.jsx:
```diff
@@ -0,0 +1,67 @@
+				</div>
+				<div className="twitter-card-preview__body">
+					<div className="twitter-card-preview__title">
+						{ title || '' }
```
These are fine to leave without the default blank. The reason I put them in above is because the truncation functions needed a string to work and I didn't want to change them. Here `{ title }` would be fine because a blank string, `undefined`, or `null` values for `title`, `description`, etc… just show up blank in React.

dmsnell commented on client/components/seo/twitter-preview/index.jsx:
```diff
@@ -0,0 +1,67 @@
+		} = this.props;
+
+		const previewClasses = classNames( 'twitter-card-preview', {
+			[ `twitter-card-preview__${ type }` ]: true
```
when these are aways true, we can add them outside of the object

```js
const previewClasses = classNames(
	'twitter-card-preview',
	`twitter-card-preview__${ type }`
);
```

Of course, as this point, it's all static so you could just put it below as a string

```js
<div className={ `twitter-card-preview twitter-card-preview__${ type }` }>
```

vindl replied:
Yes, I was wondering what was the point of that, but forgot to ask afterwards. :) 
Thanks for the advice, I'll go with the second option.

dmsnell commented on client/components/seo/twitter-preview/index.jsx:
```diff
@@ -0,0 +1,67 @@
+		return (
+			<div className={ previewClasses }>
+				<div className="twitter-card-preview__image">
+					<img className="" src={ image || blavatarUrl( slug ) } />
```
setting `className=""` isn't doing anything, we should take this out.

dmsnell commented:
looking good @vindl. I'm going to make some changes to the Facebook branch. Actually now that I think of it, I forgot this was based off of that and I have rebased, so there might be big issues with this one.

Since I don't have a local copy of this, I recommend saving a full-diff patch just in case from your computer.

```bash
git diff master... > ~/Downloads/twitter-preview-master.patch
git diff add/facebook-preview... > ~/Downloads/twitter-preview-fb-preview.patch
```

That way if something goes wrong we can recover.

vindl commented:
@dmsnell I've rebased and removed the dependency on your branch, since I've factored out those helper functions. 

I've also added horizontal scrolling on overflow in order to make it responsive and conform with Facebook preview component.

roundhill commented:
LGTM, any thoughts @dmsnell?

dmsnell commented:
Go for it!
